### PR TITLE
Feature/implement chat

### PR DESCRIPTION
**Pull request by Mmishaaa**

### CODE REVIEW

**Review by VadimZharikov — APPROVED** [review]

File: Tinder.API/Hubs/ChatHub.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            }
+        }
+
+        public async Task SendMessage(string message)
+        {
+            var stringConnection = await _cache.GetAsync(Context.ConnectionId);
+            var connection = JsonSerializer.Deserialize<ChatConnection>(stringConnection);
+
+
+            if (connection is not null)
```
Comment (moltenessence) [inline Tinder.API/Hubs/ChatHub.cs:66]: don't we need to create new connection if there is no existing one in a cache?

File: Tinder.BLL/Services/ChatService.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+using Tinder.BLL.Interfaces;
+using Tinder.BLL.Models;
+using Tinder.DAL.Entities;
+using Tinder.DAL.Interfaces;
+
+namespace Tinder.BLL.Services
+{
+    public class ChatService : GenericService<Chat, ChatEntity>, IChatService
+    {
+        private readonly IChatRepository _chatRepository;
```
Comment (moltenessence) [inline Tinder.BLL/Services/ChatService.cs:11]: formatting

File: Tinder.API/Hubs/ChatHub.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+using Tinder.Bll.Exceptions;
+using Tinder.BLL.Exceptions;
+using Tinder.BLL.Interfaces;
+using Tinder.BLL.Models;
+
+namespace Tinder.API.Hubs
+{
+    public class ChatHub : Hub<IChatHub>
+    {
+
```
Comment (moltenessence) [inline Tinder.API/Hubs/ChatHub.cs]: formatting

File: Tinder.API/Hubs/ChatHub.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            _cache = cache;
+            _mapper = mapper;
+        }
+
+        public async Task JoinChat(ChatConnection connection)
+        {
+            var chat = await _chatService.GetByIdAsync(connection.ChatId, cancellationToken: default) ?? throw new NotFoundException("Chat is not found");
+            var user = await _userService.GetByIdAsync(connection.UserId, cancellationToken: default) ?? throw new NotFoundException("User is not found");
+            
+            if (chat.Users.All(u => u.Id != user.Id)) throw new BadRequestException("User is not allowed to be in the chat");
```
Comment (moltenessence) [inline Tinder.API/Hubs/ChatHub.cs:37]: may be use !Any() here?

Comment (Mmishaaa) [reply Tinder.API/Hubs/ChatHub.cs:37]: initially i used !Any() but resharper suggested changing it to .ALL()

Comment (Mmishaaa) [reply Tinder.API/Hubs/ChatHub.cs:37]: which one is better to use or it doesn't really matter?

File: Tinder.API/Hubs/ChatHub.cs (hunk 1 above)
Comment (Mmishaaa) [reply Tinder.API/Hubs/ChatHub.cs:66]: probably, as you must join a chat before sending a message and if you join a chat you will have a new connection saved in cache, so connection cannot be null in SendMessage method and it'll be better just to remove if statement here?

Comment (moltenessence) [reply Tinder.API/Hubs/ChatHub.cs:66]: You use Redis which stores data in memory. If the Redis instance is rebooted, you lose the data in cache. Do you know what way your system going to behave in this case?

File: Tinder.API/Hubs/ChatHub.cs (hunk 4 above)
Comment (moltenessence) [reply Tinder.API/Hubs/ChatHub.cs:37]: Doesn't matter

File: Tinder.API/Hubs/ChatHub.cs (hunk 1 above)
Comment (Mmishaaa) [reply Tinder.API/Hubs/ChatHub.cs:66]: all data will be lost, won't it?

**Review by moltenessence — APPROVED** [review]

**Pull request merged**